### PR TITLE
Replace distinct with aggregate for where-in/exists subquery

### PR DESCRIPTION
**Pull request by ygf11**
# Which issue does this PR close?

Part of #5429.

# Rationale for this change

# What changes are included in this PR?

1. Find `where-in/exists` and optimize them.
2. Support Aggregate subquery for where-exists to join, when the aggregate comes from distinct.  

# Are these changes tested?
Yes.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by ygf11** [comment]
Please take a look @mingmwang @jackwener @alamb

File: datafusion/optimizer/src/decorrelate_where_exists.rs [hunk 1]
```diff
@@ -670,4 +678,56 @@ mod tests {
...
+                        \n        Projection: sq.a, sq.c [a:UInt32, c:UInt32]\
+                        \n          Filter: sq.a > test.b [a:UInt32, b:UInt32, c:UInt32]\
+                        \n            TableScan: sq [a:UInt32, b:UInt32, c:UInt32]\
+                        \n    TableScan: test [a:UInt32, b:UInt32, c:UInt32]";
+
+        assert_plan_eq(&plan, expected)
+    }
+
+    #[test]
+    fn exists_subquery_aggragte_distinct() -> Result<()> {
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_where_exists.rs:711]:
```suggestion
    fn exists_subquery_aggregate_distinct() -> Result<()> {
```

File: datafusion/optimizer/src/decorrelate_where_exists.rs [hunk 2]
```diff
@@ -670,4 +678,56 @@ mod tests {
...
+        let table_scan = test_table_scan()?;
+        let subquery_scan = test_table_scan_with_name("sq")?;
+
+        let subquery = LogicalPlanBuilder::from(subquery_scan)
+            .filter(col("sq.a").gt(col("test.b")))?
+            .project(vec![col("sq.a"), col("sq.c")])?
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_where_exists.rs:717]: Can you add a test for:
1. When there is no projection in the subquery?
2. When the projection in the subquery is an expression (like `sq.a + sq.b`)?

File: datafusion/optimizer/src/replace_distinct_aggregate.rs [hunk 3]
```diff
@@ -100,4 +173,164 @@ mod tests {
...
+
+    #[test]
+    fn replace_distinct_in_where_in() -> datafusion_common::Result<()> {
+        let table_scan = test_table_scan()?;
+        let subquery_scan = test_table_scan_with_name("sq")?;
+
+        // distinct in where-in subquery
+        let subquery = LogicalPlanBuilder::from(subquery_scan)
+            .filter(col("test.a").eq(col("sq.a")))?
+            .project(vec![col("sq.b"), col("sq.c")])?
```
Comment (alamb) [inline datafusion/optimizer/src/replace_distinct_aggregate.rs:218]: In these tests too I recommend projecting an expression (not just columns)

**Review by alamb — COMMENTED** [review]
Thanks @ygf11  -- I think this code looks quite good

I am not sure if this will work with expressions (not just column references) but a few more tests I think will demonstrate one way or the other


cc @mingmwang

File: datafusion/expr/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -1776,6 +1776,40 @@ impl Aggregate {
             _ => plan_err!("Could not coerce into Aggregate!"),
         }
     }
+
+    /// Check whether it is a Distinct.
+    /// A Distinct means all fields of the schema are the expressions of group by.
```
[the remark this answers is not in the document]
Comment (mingmwang) [reply datafusion/expr/src/logical_plan/plan.rs:1781]: I also feel confusing to the naming and the implementation of this method. Maybe we should call this is_group_by_only()

**Comment by mingmwang** [comment]
@ygf11 @jackwener 
I'm not sure whether we should make the rule `ReplaceDistinctWithAggregate` to handle the where-in/exists subquery specifically. My original thinking was let the rules `DecorrelateWhereExists` and `DecorrelateWhereIn`  rewrite the subqueries to Joins and in the second pass the rule `ReplaceDistinctWithAggregate`  will rewrite the `Distinct` to `Aggregate` since we already run those rules multiple times

**Comment by ygf11** [comment]
> I'm not sure whether we should make the rule ReplaceDistinctWithAggregate to handle the where-in/exists subquery specifically. My original thinking was let the rules DecorrelateWhereExists and DecorrelateWhereIn rewrite the subqueries to Joins and in the second pass the rule ReplaceDistinctWithAggregate will rewrite the Distinct to Aggregate since we already run those rules multiple times, so that we can keep a relatively simple ReplaceDistinctWithAggregate rule.
For those expr subqueries that can not be decorrelated, we can create another optimization task and apply all the existing rules to them

Thanks @mingmwang. One thing I care about If we rewrite distinct in second pass, is that we have to run the rules one more time sometimes, if it is not a problem, it is ok to me.

**Comment by jackwener** [comment]
> @ygf11 @jackwener I'm not sure whether we should make the rule `ReplaceDistinctWithAggregate` to handle the where-in/exists subquery specifically. My original thinking was let the rules `DecorrelateWhereExists` and `DecorrelateWhereIn` rewrite the subqueries to Joins and in the second pass the rule `ReplaceDistinctWithAggregate` will rewrite the `Distinct` to `Aggregate` since we already run those rules multiple times, so that we can keep a relatively simple `ReplaceDistinctWithAggregate` rule.
> 
> For those expr subqueries that can not be decorrelated, we can create another optimization task and apply all the existing rules to them

Agree with it

**Comment by mingmwang** [comment]
Yes, I think most of the rules do not need to handle subqueries specifically,  we need to apply PushDownFilter, PushDownLimit... to  subqueries also, if the subqueries can not be decorrelated. we can have another optimization task/process and apply the existing rules to subqueries.

There is another PR related to subqueries and distinct, I think let's get this PR merge first.
https://github.com/apache/arrow-datafusion/pull/5345

**Comment by mingmwang** [comment]
@ygf11 
Can we close this PR now as this is not required ?.

**Pull request closed without merging**